## Упражнения по библиотеке Numpy

In [2]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [ ]:
a = np.arange(9)
for i in a:
    if 3 < i < 8:
        a[i] = i * -1
print(a)

[ 0  1  2  3 -4 -5 -6 -7  8]


**2.** Заменить максимальный элемент случайного массива на 0

In [ ]:
a = np.random.random(10)
a[a.argmax()] = 0
print(a)

[0.81467737 0.44123504 0.3440192  0.28865571 0.49761912 0.15889325
 0.69780489 0.         0.28197602 0.36048207]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [ ]:
import itertools as it

array = np.array([1, 2, 3])

combinations = it.product(array,array)

for combination in combinations:
    print(combination)

(1, 1)
(1, 2)
(1, 3)
(2, 1)
(2, 2)
(2, 3)
(3, 1)
(3, 2)
(3, 3)


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [ ]:
a = np.random.randint(0,5,(8,3))
b = np.random.randint(0,5,(2,2))
c = (a[..., np.newaxis, np.newaxis] == b)
rows = np.where(c.any((3,1)).all(1))[0]
print(rows)

[1 4 5 6]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [ ]:
a = np.random.randint(0,5,(10,3))
b = a[a.max(axis = 1) != a.min(axis = 1),:]
print(b)

[[4 0 1]
 [3 2 0]
 [0 4 0]
 [4 2 0]
 [2 1 2]
 [1 3 2]
 [1 0 4]
 [2 1 4]
 [4 1 1]
 [0 1 2]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [ ]:
a = [[1, 0, 1, 0, 0, 1],
     [1, 1, 1, 0, 0, 1],
     [1, 0, 1, 0, 0, 1],
     [1, 0, 1, 0, 1, 1],
     [1, 0, 1, 0, 0, 1]]

b = np.array(a)
np.unique(b, axis = 0)

array([[1, 0, 1, 0, 0, 1],
       [1, 0, 1, 0, 1, 1],
       [1, 1, 1, 0, 0, 1]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [ ]:
import math
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
def prod_non_zero_diag_non_vect(X):
    len_diag = min(len(X), len(X[0]))
    mult = 1
    for i in range(len_diag):
        if(X[i][i] != 0):
            mult *= X[i][i]
    return mult
prod_non_zero_diag_non_vect(X)

3

In [ ]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
def prod_non_zero_diag_vect(X):
    diag = np.diag(X)
    return diag[diag != 0].prod()
prod_non_zero_diag_vect(X)

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [ ]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
def are_multisets_equal_non_vect(x, y):
    if(len(x) != len(y)):
        return False
    x.sort()
    y.sort()
    for i in range(len(x)):
        if(x[i] != y[i]):
            return False
    return True
are_multisets_equal_non_vect(x, y)

True

In [ ]:
x = ([1, 2, 2, 4])
y = ([4, 2, 1, 2])
def are_multisets_equal_vect(x, y):
    x_nums, x_counts = np.unique(x, return_counts=True)
    y_nums, y_counts = np.unique(y, return_counts=True)
    if(np.shape(x_nums) != np.shape(y_nums)):
        return False
    if(np.any(x_nums != y_nums) or np.any(x_counts != y_counts)):
        return False
    return True
are_multisets_equal_vect(x, y)

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [ ]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
def max_after_zero_non_vect(x):
    after_zero_list = list()
    for i in range(1, len(x)):
        if(not x[i - 1]):
            after_zero_list.append(x[i])
    return max(after_zero_list)
max_after_zero_non_vect(x)

5

In [ ]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
def max_after_zero_vect(x):
    curr = np.ones((1))
    curr = np.hstack((curr, x))
    ind = curr == 0
    return np.max(x[ind[:-1]])
max_after_zero_vect(x)

5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [ ]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
def run_length_encoding_non_vect(x):
    val_list = [x[0]]
    num_list = list()
    counter = 1
    for i in range(1, len(x)):
        if(x[i - 1] == x[i]):
            counter = counter + 1
        else:
            val_list.append(x[i])
            num_list.append(counter)
            counter = 1
    num_list.append(counter)
    return (val_list, num_list)
run_length_encoding_non_vect(x)

([2, 3, 5], [3, 3, 1])

In [ ]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
def run_length_encoding_vect(x):
    y = np.hstack((np.ones(1), x[:- 1]))
    first_positions = x != y
    first_positions[0] = True
    indexes_1 = np.arange(np.size(x))[first_positions]
    indexes_2 = np.hstack((indexes_1[1:], np.array([np.size(x)])))
    return x[first_positions], indexes_2 - indexes_1
run_length_encoding_vect(x)

(array([2, 3, 5]), array([3, 3, 1]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [ ]:
X = np.array(
    [[243,  3173],
     [525,  2997]])
Y = np.array(
    [[682, 2644],
     [277, 2651],
     [396, 2640]])
def pairwise_distance_non_vect(X, Y):
    result_matr = list()
    for i in range(len(X)):
        curr_list = list()
        for j in range(len(Y)):
            dist = 0
            for k in range(len(X[0])):
                dist += (X[i][k] - Y[j][k]) ** 2
            curr_list.append(math.sqrt(dist))
        result_matr.append(curr_list)
    return result_matr
pairwise_distance_non_vect(X, Y)
%timeit pairwise_distance_non_vect(X, Y)

100000 loops, best of 5: 17.1 µs per loop


In [ ]:
X = np.array(
    [[243,  3173],
     [525,  2997]])
Y = np.array(
    [[682, 2644],
     [277, 2651],
     [396, 2640]])
def pairwise_distance_vect(X, Y):
    return np.sqrt(np.sum((X[:, np.newaxis] - Y) ** 2, axis=-1))
pairwise_distance_vect(X, Y)
%timeit pairwise_distance_vect(X, Y)

The slowest run took 8.17 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 10.9 µs per loop


In [9]:
from scipy import spatial
X = np.array(
    [[243,  3173],
     [525,  2997]])
Y = np.array(
    [[682, 2644],
     [277, 2651],
     [396, 2640]])
def pairwise_distance_scipy(X, Y):
  return np.min(spatial.distance.cdist(X, Y), axis = 1)
pairwise_distance_scipy(X, Y)
%timeit pairwise_distance_scipy(X, Y)

The slowest run took 17.37 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 5: 17.3 µs per loop


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [121]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [141]:
import numpy as np
calorie_stats = np.loadtxt("cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [164]:
calories_mean = np.mean(calorie_stats)
average_calories = calories_mean - 60
calories_mean
average_calories

46.883116883116884

In [165]:
calories_mean

106.88311688311688

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [148]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [163]:
median_calories = np.median(calorie_stats)
print('median_calories:',median_calories,'- calories_mean:',np.round(np.mean(calorie_stats),1))

median_calories: 110.0 - calories_mean: 106.9


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [168]:
np.percentile(calorie_stats, 1)
np.percentile(calorie_stats, 2)
np.percentile(calorie_stats, 3)
np.percentile(calorie_stats, 4)



70.0

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [160]:
more_calories = np.mean(calorie_stats > 60)
more_calories

0.961038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [169]:
calorie_std = np.std(calorie_stats)
calorie_std

19.35718533390827

Это стандартное отклонение, которое поможет нам понять как распределены значения относительно среднего значения в нашей выборке. Чем ближе стандартное отклонение к 0, тем надежнее среднее.

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Было проанализировано 77 различных марок хлопьев. В 96% из анализируемых хлопьев калорийнось выше, чем у CrunchieMunchies. B среднем калорийность составляла 106,9 калорий со стандартным отклонением 19,4. Лишь 3 марки имеют калорийность меньше, чем CrunchieMunchies.